pkg/grpc/api_local.go

```diff
@@ -0,0 +1,232 @@
+	"github.com/dapr/dapr/pkg/messages"
+	"github.com/dapr/dapr/pkg/messaging"
+	invokev1 "github.com/dapr/dapr/pkg/messaging/v1"
+	commonv1pb "github.com/dapr/dapr/pkg/proto/common/v1"
+	internalv1pb "github.com/dapr/dapr/pkg/proto/internals/v1"
+)
+
+// Maximum size for the buffer used by CallLocalStream.
+// Equivalent to 4KB.
+const StreamBufferSize = 4096
+
+// CallLocal is used for internal dapr to dapr calls. It is invoked by another Dapr instance with a request to the local app.
+func (a *api) CallLocal(ctx context.Context, in *internalv1pb.InternalInvokeRequest) (*internalv1pb.InternalInvokeResponse, error) {
+	if a.appChannel == nil {
+		return nil, status.Error(codes.Internal, messages.ErrChannelNotFound)
+	}
+
+	req, err := invokev1.InternalInvokeRequest(in)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, messages.ErrInternalInvokeRequest, err.Error())
+	}
+	defer req.Close()
+
+	err = a.callLocalACL(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := a.appChannel.InvokeMethod(ctx, req)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, messages.ErrChannelInvoke, err)
+	}
+	defer resp.Close()
+
+	// Response message
+	return resp.ProtoWithData()
+}
+
+// CallLocalStream is a variant of CallLocal that uses gRPC streams to send data in chunks, rather than in an unary RPC.
+// It is invoked by another Dapr instance with a request to the local app.
+func (a *api) CallLocalStream(stream internalv1pb.ServiceInvocation_CallLocalStreamServer) error { //nolint:nosnakecase
+	if a.appChannel == nil {
+		return status.Error(codes.Internal, messages.ErrChannelNotFound)
+	}
+
+	ctx, cancel := context.WithCancel(stream.Context())
+	defer cancel()
+
+	// Read the first chunk of the incoming request
+	// This contains the metadata of the request
+	chunk := &internalv1pb.InternalInvokeRequestStream{}
+	err := stream.RecvMsg(chunk)
+	if err != nil {
+		return err
+	}
+	if chunk.Request == nil || chunk.Request.Metadata == nil || chunk.Request.Message == nil {
+		return status.Errorf(codes.InvalidArgument, messages.ErrInternalInvokeRequest, "request does not contain the required fields in the leading chunk")
+	}
+	pr, pw := io.Pipe()
+	req, err := invokev1.InternalInvokeRequest(chunk.Request)
+	if err != nil {
+		return status.Errorf(codes.InvalidArgument, messages.ErrInternalInvokeRequest, err.Error())
+	}
+	req.WithRawData(pr, chunk.Request.Message.ContentType)
+	defer req.Close()
+
+	err = a.callLocalACL(ctx, req)
+	if err != nil {
+		return err
+	}
+
+	// Read the rest of the data in background as we submit the request
+	go func() {
+		var (
+			done    bool
+			readErr error
+		)
+		for {
+			if ctx.Err() != nil {
+				pw.CloseWithError(ctx.Err())
+				return
+			}
+
+			done, readErr = messaging.ReadChunk(chunk, pw)
+			if readErr != nil {
+				pw.CloseWithError(readErr)
+				return
+			}
+
+			if done {
+				break
+			}
+
+			readErr = stream.RecvMsg(chunk)
+			if err != nil {
```

Comment:
Should this not reference readErr?

Reply:
Yes it should be. Thanks for catching it